src/pip/_internal/vcs/versioncontrol.py

```diff
@@ -625,10 +656,17 @@ def controls_location(cls, location):
         # type: (str) -> bool
         """
         Check if a location is controlled by the vcs.
-        It is meant to be overridden to implement smarter detection
-        mechanisms for specific vcs.
 
-        This can do more than is_repository_directory() alone.  For example,
-        the Git override checks that Git is actually available.
+        Searches up the filesystem and checks is_repository_directory().
+
+        It is meant to be extended to add smarter detection mechanisms for
+        specific vcs.  For example, the Git override checks that Git is
+        actually available.
         """
-        return cls.is_repository_directory(location)
+        while not cls.is_repository_directory(location):
```

Comment:
I think this behavior change is unrelated to the issue?

Reply:
This was a halfway measure towards implementing #7162.  Will revert it back.